### PR TITLE
Derive elliptic solutions from base classes

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

The classes use the registration mechanism so the executables can choose
which set of classes they want to use. Because of the extra template
parameter the implementations of the solution variables have to move to
a helper class so they can be template-instantiated in a cpp file. This
has the benefit that it encourages using the CachedTempBuffer data
structure, which helps with re-using intermediate quantities.

(I don't actually refactor the implementations to make great use of the
CachedTempBuffer in this PR, that can be done later if needed)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@fmahebert and @nilsdeppe would you be able to review this one? It was split out of #2603 a while ago

File: src/PointwiseFunctions/AnalyticData/AnalyticData.hpp [hunk 1]
```diff
@@ -3,7 +3,53 @@
...
+#include "Utilities/Registration.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
 /// \cond
 // Empty base class for marking analytic data.
 struct MarkAsAnalyticData {};
 /// \endcond
+
+/// Provides analytic tensor data as a function of the spatial coordinates
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/AnalyticData.hpp]: I suggest to add a little "\note" to this Doxygen to clarify..
1. There are now two base classes in this file. When should I use which base class? Is this new one for use primarily in the elliptic code?
2. The AnalyticSolutions for evolution problems are marked as AnalyticSolutions only, and not also as AnalyticData, but here it seems you are deriving the elliptic AnalyticSolution from the AnalyticData base class. Can you add a note to document this design?

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp [hunk 2]
```diff
@@ -4,26 +4,60 @@
...
+struct BentBeamVariables {
+  using Cache = CachedTempBuffer<BentBeamVariables, Tags::Displacement<2>,
+                                 Tags::Strain<2>, Tags::Stress<2>,
+                                 ::Tags::FixedSource<Tags::Displacement<2>>>;
+
+  const tnsr::I<DataType, 2>& x;
+  const double length_;
+  const double height_;
+  const double bending_moment_;
+  const ConstitutiveRelations::IsotropicHomogeneous<2>& constitutive_relation_;
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.hpp]: Stylistic nitpick: I believe we use trailing underscores on private members, so these should have no trailing underscore.

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/HalfSpaceMirror.hpp [hunk 3]
```diff
@@ -6,23 +6,57 @@
...
+  using Cache = CachedTempBuffer<HalfSpaceMirrorVariables,
+                                 Tags::Displacement<3>, Tags::Strain<3>,
+                                 ::Tags::FixedSource<Tags::Displacement<3>>>;
+
+  const tnsr::I<DataType, 3>& x;
+  const double beam_width_;
+  const ConstitutiveRelations::IsotropicHomogeneous<3>& constitutive_relation_;
+  const size_t integration_intervals_;
+  const double absolute_tolerance_;
+  const double relative_tolerance_;
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticSolutions/Elasticity/HalfSpaceMirror.hpp]: [same stylistic nitpick]

File: src/PointwiseFunctions/AnalyticData/AnalyticData.hpp (hunk 1 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/AnalyticData.hpp]: Added some more docs to address both points. I expect the `MarkAsAnalyticData` struct will be deleted once all executables factory-create the analytic data classes.

**Comment by nilsvu** [comment]
Thanks for reviewing @fmahebert! I pushed fixups, please take a look when you get the chance.

@nilsdeppe or @wthrowe could you do the second review?

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/AnalyticSolution.hpp [hunk 4]
```diff
@@ -17,7 +17,18 @@ namespace Elasticity {
...
+ * \brief Base class for analytic solutions of the linear Elasticity equations
+ *
+ * This abstract base class allows factory-creating derived classes from
+ * input-file options (see `Registration`). The list of `Registrars` may also
+ * contain classes that aren't analytic solutions but only derive off
+ * `Elasticity::AnalyticData::AnalyticData`. This allows factory-creating any
+ * choice of analytic solution _or_ analytic data from input-file options. Note
+ * that analytic solutions provide more variables than analytic data classes, so
+ * to request variables that only analytic solutions provide you must perform a
+ * cast to `AnalyticSolution`.
```
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/Elasticity/AnalyticSolution.hpp]: This doesn't make sense.  You can't factory create something that isn't an AnalyticSoltuon using an AnalyticSoluton factory.  (And again in another file.)

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/AnalyticSolution.hpp (hunk 4 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Elasticity/AnalyticSolution.hpp]: You're right, reworded the docs slightly

**Comment by nilsvu** [comment]
@wthrowe reworded the docs in the fixup, please take another look. Cheers!

**Comment by wthrowe** [comment]
OK, I think I understand now.  You don't actually want an AnalyticSolution factory, the registrars are serving the same purpose they do in the most-derived classes.

In that case, the AnalyticSolution class should not define a `creatable_classes` type alias.  (Although, if I remember the rules correctly, it will inherit the one from AnalyicData, so this is more documentation than a functional change to the class.)

**Comment by nilsvu** [comment]
@wthrowe I kept the `creatable_classes` alias in `AnalyticSolution` so the class can be used for factory-creation in the case where _all_ registrars are analytic solutions. Does that make sense?

**Comment by wthrowe** [comment]
OK, I see.  Let's stick with your most recent edits, then.  Go ahead and squash.

**Comment by nilsvu** [comment]
Rebased and squashed. Thanks for your reviews @fmahebert and @wthrowe!

**Review by wthrowe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Pull request merged**